src/json2typescript/json-convert.ts

```diff
@@ -228,7 +228,7 @@ export class JsonConvert {
      *
      * @see https://www.npmjs.com/package/json2typescript full documentation
      */
-    serialize<T>(data: any | any[], classReference?: { new(): T }): any | any[] {
+    serialize<T extends object, U extends object = {}>(data: T | T[], classReference?: { new(): U }): object | object[] {
```

Comment:
We want to  use a different generic type because we now allow that the data has actually a different structure that the class reference. I use the return type object because it should be a plain JS object now.

Reply:
I don't understand `U extends object = {}`. Doesn't this mix a type and an instance?

I guess what you want to say is that U is of type object (a subclass), but why create an empty object? That should be handled by the implementation of the method.

Reply:
or is this like a type annotation for the return type of an arrow function?

Reply:
see https://www.typescriptlang.org/docs/handbook/generics.html#generic-types

Reply:
The second parameter of type U is optional. Afaik this is the way to let the compiler know that U belongs to an optional parameter and must not be passed (https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-3.html#generic-parameter-defaults):

Without default generic value, you could not write `serialize<User>(user)`, you would have to write `serialize<User, User>(user)`, even though the second generic is useless.

Reply:
I think in general, if the type can be inferred, you do not have to pass the type also if there is no default:

> `let output = identity("myString");  // type of output will be 'string'`
Notice that we didn’t have to explicitly pass the type in the angle brackets (<>); the compiler just looked at the value "myString", and set T to its type.  While type argument inference can be a helpful tool to keep code shorter and more readable, you may need to explicitly pass in the type arguments as we did in the previous example when the compiler fails to infer the type, as may happen in more complex examples.
https://www.typescriptlang.org/docs/handbook/generics.html#hello-world-of-generics

But in this case, I think the compiler cannot figure it out because it is the return type.
So the following applies:
> If a default type is specified and inference cannot choose a candidate, the default type is inferred.
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-3.html#generic-parameter-defaults

In short, I think what you do is correct. :-)